datalad_next/tree.py

```diff
@@ -0,0 +1,1172 @@
+    if the symlink and its target are both included in the tree.
+
+    Results are cached because the check is somewhat expensive and may
+    be run multiple times on the same path.
+
+    TODO: is there a way to detect a datalad dataset if it is not installed
+     and it is not a subdataset?
+
+    Parameters
+    ----------
+    path: Path
+        Path to directory to be identified as dataset or non-dataset
+    """
+    try:
+        if path.is_symlink():
+            # ignore symlinks even if pointing to datasets, otherwise we may
+            # get duplicate counts of datasets
+            lgr.debug("Path is a symlink, will not check if it points to a "
+                      f"dataset: '{path}'")
+            return False
+
+        if (path / ".datalad" / "config").is_file() or \
+                (path / ".datalad" / "metadata").is_dir():
+            # could also query `ds.id`, but checking just for existence
+            # of config file is quicker.
+            return True
+
+        # if it is not installed, check if it has an installed superdataset.
+        # instead of querying ds.is_installed() (which checks if the
+        # directory has the .git folder), we check if the directory
+        # is empty (faster) -- as e.g. after a non-recursive `datalad clone`
+        if is_empty_dir(path):
+            if get_superdataset(path) is not None:
+                return True
+
+    except Exception as ex:
+        # if anything fails (e.g. permission denied), we raise exception
+        # instead of returning False. this can be caught and handled by the
+        # caller.
+        raise NoDatasetFound(f"Cannot determine if '{path.name}' is a "
+                             f"dataset") from ex
+
+    return False
+
+
+@lru_cache
+def get_subds_paths(ds_path: Path):
+    """Return paths of immediate subdatasets for a given dataset path."""
+    # This is an expensive operation because it calls git to read the
+    # submodules. Since we need to run it to (A) calculate dataset depth and
+    # (B) detect non-installed datasets, we cache results, so that the list of
+    # subdatasets is computed only once for each parent dataset.
+
+    def res_filter(res):
+        return res.get('status') == 'ok' and res.get('type') == 'dataset'
+
+    # call subdatasets command instead of dataset method `ds.subdatasets()`
+    # to avoid potentially expensive import of full datalad API
+    return Subdatasets.__call__(
+        dataset=ds_path,
+        recursive=False,
+        state='any',  # include not-installed subdatasets
+        result_filter=res_filter,
+        on_failure='ignore',
+        result_xfm='paths',
+        result_renderer='disabled',
+        return_type='list'
+    )
+
+
+def is_subds_of_parent(subds_path: Path, parent_path: Path):
+    return any(
+        is_path_relative_to(Path(p), subds_path)
+        for p in get_subds_paths(parent_path)
+    )
+
+
+def get_dataset_root_datalad_only(path: Path):
```

Comment:
```suggestion
@lru_cache
def get_dataset_root_datalad_only(path: Path):
```

On my system and the 80k directory dataset, this gives 10-15% speedup and has no performance impact in for directories not contained in a dataset.

Reply:
Makes sense! Fixed in af014207.